### DISCUSSION TITLE
How to create an API rule which checks if the request user is valid.

### DISCUSSION THREAD

**Discussion by Frenzoid** [pocketbase/pocketbase#1379] [Q&A] [answered]
Let's say i have two collections, users ( the default user collection ), and messages, and I only want to allow registered users to create new messages.

I've seen from the docs that this rule on CREATE ACTION is enough: `@request.auth.id != ""`.

![image](https://user-images.githubusercontent.com/30338644/209467890-f7a682fb-a69b-46c8-8f19-bac4f03ec67e.png)

But wouldn't this allow any user with any string as an id to create new messages?
Also, I want to ensure that the request has a valid session token since I don't want other people mimicking other user's ids on their request.

Best regards!

**Comment by ganigeorgiev** [comment, marked as answer]
You can define a `createdBy`, `user`, `author` etc. **nonempty** `relation` field in your `messages` collection and use the following rule for create, update and delete:

```js
// create
@request.data.author = @request.auth.id

// allow update only by the message author
author = @request.auth.id && (@request.data.author = "" || @request.data.author = @request.auth.id)

// allow delete only by the message author
author = @request.auth.id
```

Note that in the above **update** rule we also allow empty "author" data submission because the request is PATCH and if `author` is missing from the request body it will not update it (if the user explicitly set an empty id it will trigger the nonempty validation). In the upcoming v0.11 release this will be a little more clear since we'll have `isset` `@request.data.*` filter modifier and the update rule could look like this:

```js
// allow update only by the message author and don't allow author change
author = @request.auth.id && @request.data.author:isset = false
```

> Also, I want to ensure that the request has a valid session token since I don't want other people mimicking other user's ids on their request.

If `@request.auth.id` is nonempty, this means that the request was made with valid `Authorization` header token and you don't have to worry about "mimicking" other users.

_edit: Updated the `isset` example to reflect the most up to date version._

**Reply by realJogicodes** [reply]
Has how this is handled changed since this post? 

I can make the  create rule
`@request.data.author = @request.auth.id`

as it is described above but when I try do make a delete rule:

`author = @request.auth.id`
Pocketbase returns:
`Invalid filter rule. Raw error: invalid left operand "author" - unknown field "author".`

Or any other rule with `author = ...`

**Reply by ganigeorgiev** [reply]
@proofofjogi `author` in the above refers to a field in an example/imaginary collection. If you don't have an `author` field then the "unknown field" error is expected. If you need further help, please open a new discussion with more details about what you are trying to do and ideally export of your collections structure as json (_Admin UI > Settings > Export collections_).

**Reply by realJogicodes** [reply]
Thank you, I might be able to fix with that. The clue was in the message which makes sense now after your explanation, I appreciate it.

Edit: I could figure it out, thanks @ganigeorgiev . `author` in the original post is the field name for the backref, which in my case stayed `field1` after adding in the relation. I forgot to change the name from that default to `user`. Changing it fixed it.

**Comment by Frenzoid** [comment]
Your answer was quick, informative, and on point. Thanks!
